Bricks/sync/owned_borrowed.h

```diff
@@ -0,0 +1,492 @@
+//   The basic usage of `master` is `master->MemberFunctionOfT(...)`, much like with C++'s `shared_ptr<T>`.
+//
+// * Borrowed<T> borrower(master);  // Or `Borrowed<T> another_borrower(borrower);`.
+//   Initializes `borrower`, which can be used same way as `master`.
+//   Casting it to `bool` (or just checking for `if (!borrower) { ... }`) would result in `false`
+//   if and only if the master scope (the scope of the `Owned<T>` is being terminated, and is currently
+//   awaiting until all the borrowers are done.
+//
+// * BorrowedWithCallback<T> borrower(master, [&signal]() { signal.SignalTermination(); });
+//   Initializes `borrower` passing it an external signal that will be likely `std::move()` it into a different thread.
+//   The lifetime of the `borrower` object will be strictly inside the lifetime of `master`.
+//   The destructor of `master` will wait until all the `borrower`-s have terminated.
+//   The callbacks, corresponding to active `borrower`-s, will be called as `master` will be destructing.
+
+#ifndef BRICKS_SYNC_OWNED_BORROWED_H
+#define BRICKS_SYNC_OWNED_BORROWED_H
+
+#include "../../port.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <vector>
+
+#include "../exception.h"
+
+#include "../template/decay.h"
+
+namespace current {
+namespace sync {
+
+// The container for the actual instance and its environment.
+namespace impl {
+
+// Helper to prohibit copying `Owned`.
+struct ConstructOwned {};
+struct ConstructBorrowable {};
+struct ConstructBorrowedObjectViaMoveConstructor {};
+
+struct ConstructUniqueContainerViaMoveConstructor {};
+
+// The actual instance, kept along with its destructing status, and a list of registered borrowers.
+template <typename T>
+struct UniqueInstance final {
+  // Constructor: Construct an instance of the object.
+  UniqueInstance() : destructing_(false), instance_(), total_borrowers_spawned_throughout_lifetime_(0u) {}
+  template <typename... ARGS>
+  UniqueInstance(ARGS&&... args)
+      : destructing_(false), instance_(std::forward<ARGS>(args)...), total_borrowers_spawned_throughout_lifetime_(0u) {}
+
+  // Destructor: Block until all the borrowers are done.
+  ~UniqueInstance() {
+    // Mark the instance of the object as being destructed.
+    destructing_ = true;
+
+    // Collect all the "destructing" callbacks to call.
+    std::vector<std::function<void()>> callbacks_to_call;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      for (const auto& cit : borrowers_) {
+        callbacks_to_call.push_back(cit.second);
+      }
+    }
+
+    // Call all the "destructing" callbacks.
+    for (auto& f : callbacks_to_call) {
+      f();
+    }
+
+    // Wait until all the borrowers have terminated.
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      condition_variable_.wait(lock, [this]() { return borrowers_.empty(); });
```

Comment:
Great idea. I'm also hoping to eventually use the `BorrowedOfGuaranteedLifetime`.

Could you remind me next time we chat?